### ISSUE TITLE
ipv6 Only Servers

### ISSUE THREAD

**Issue by eric1234** [basecamp/kamal#588]
I'm testing out deploying and running an app on an ipv6 only server with that deployment happening via Kamal. I see on https://github.com/basecamp/kamal/issues/87 that ipv6 should work but I've encountered a few issues.

I have some thoughts about fixes but wasn't sure of them so I figured I would start with opening an issue. If I can get a bit of feedback on it I would be glad to do the legwork to help resolve these issues. So far there are three issues.

Mismatch Between Host and Configuration String
-------------------------------------------------------------------------

This prevents the deploy from working as the env files won't push up (may affect other commands but since it fails so early not sure what else). The issue is when I specify my ipv6 address I must include the brackets around the address (similar to what web browsers require). Example:

```yaml
servers:
  - "[2a01:4f7:c014:160::1]"
```

You can see these brackets being required in SSHKit [here](https://github.com/capistrano/sshkit/blob/3555a98d5c61e26472e988b62601fd323b374176/lib/sshkit/host.rb#L154)

The problem is both this line:

https://github.com/basecamp/kamal/blob/635876bdb96eb00ed85c91fe7231e05a25fd16ad/lib/kamal/commander.rb#L59

and this line:

https://github.com/basecamp/kamal/blob/635876bdb96eb00ed85c91fe7231e05a25fd16ad/lib/kamal/commander.rb#L75

assume the configuration and hostname are the same. But this is really only true for the [SimpleHostParser](https://github.com/capistrano/sshkit/blob/3555a98d5c61e26472e988b62601fd323b374176/lib/sshkit/host.rb#L108-L110)

[All other host parsers](https://github.com/capistrano/sshkit/blob/3555a98d5c61e26472e988b62601fd323b374176/lib/sshkit/host.rb#L188-L191) do something with that initialization string. Therefore when the two problem lines call `to_s` they are only getting the hostname (which does not include the brackets) while the values it's comparing against from the config file do include the brackets.

The additional challenge here is that `host` isn't always a `SSHKit::Host` object. When doing an interactive execution it just becomes a string derived from `Kamal.primary_host` at:

https://github.com/basecamp/kamal/blob/635876bdb96eb00ed85c91fe7231e05a25fd16ad/lib/kamal/cli/app.rb#L126

With a simple host, just calling `to_s` on either type of object works (`String` or `SSHKit::Host`). But for hosts parsed by any other parser `to_s` is just the hostname which is incomplete. A bit of a hacky solution I implemented to move past it was to monkey-patch `to_s` in SSHKit with:

```ruby
class SSHKit::Host
  module StringifyToFullConfig
    def initialize(configuration)
      @configuration = configuration

      super
    end

    def to_s = @configuration
  end

  prepend StringifyToFullConfig
end
```

On one hand this seems "good enough" as it works. OTOH what downstream effects will be cause by overwriting `to_s` in a 3rd party package?

Interactive SSH Host Format Invalid
---------------------------------------------------

The next issue is when I try to do an interactive ssh session I end up with the reverse problem. The `ssh` CLI tool needs it without the brackets and if it has brackets then it assumes a hostname that it tries to lookup in the DNS. Example:

```sh
$ kamal app exec -i bash
Get most recent version available as an image...
Launching interactive command with version latest via SSH from new container on [2a01:4f7:c014:160::1]...
ssh: Could not resolve hostname [2a01:4f7:c014:160::1]: Name or service not known
```

In this case we want to have the connection string parsed out and just provide the `ssh` program the hostname. I resolved this by changing the `run_over_ssh` so it converts the host to a `SSHKit::Host` then extracts just the hostname:

```diff
def run_over_ssh(*command, host:)
+  host = SSHKit::Host.new host

  "ssh".tap do |cmd|
    if config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Jump)
      cmd << " -J #{config.ssh.proxy.jump_proxies}"
    elsif config.ssh.proxy && config.ssh.proxy.is_a?(Net::SSH::Proxy::Command)
      cmd << " -o ProxyCommand='#{config.ssh.proxy.command_line_template}'"
    end
-    cmd << " -t #{config.ssh.user}@#{host} -p #{config.ssh.port} '#{command.join(" ")}'"
+    cmd << " -t #{config.ssh.user}@#{host.hostname} -p #{config.ssh.port} '#{command.join(" ")}'"
  end
end
```

With that change in place I'm able to connect just fine into my container.

Configure Docker for ipv6
--------------------------------------

My final issue (so far!), is that Docker by doesn't setup its default bridge network with ipv6 support by default. They [provide instructions](https://docs.docker.com/config/daemon/ipv6/#use-ipv6-for-the-default-bridge-network) that are straight-forward. Just drop in a config file, restart docker and your good to go. Now containers being run can call other ipv6 services (such as the database server). The result is to setup I really need to:

1. Run `kamal server bootstrap` to install docker
2. Manually drop that file and restart docker on all servers
3. Run `kamal setup` to finish the rest of the setup

It would be nice if I could just run `kamal setup` and it would be ipv6 ready. This starts to seem like a Docker problem. But OTOH Kamal is taking responsibility for installing Docker. So perhaps it should also enable ipv6?

It could maybe do this unconditionally (even if not using ipv6). I don't think it hurts anything. Or perhaps we could pass a `-ipv6` switch with `kamal setup` if we didn't want it global. Another solution might be a hook. Just run a hook on the server before being bootstraped which would allow me to have a shell script that drops the config in the right place.

**Comment by zealot128** [comment]
I also stumbled upon this issue and wondered why no env was pushed to any server without any message, even after wrapping the IP in brackets. 
(Missing the brackets just crashes with:
``ERROR (SocketError): Exception while executing on host 2a01: getaddrinfo: Temporary failure in name resolution``)

Using IPv6 one could remove the IPv4 address from all containers and save some bucks in the public cloud, like $3-4 on AWS, and $0,5 on Hetzner, and also significantly hide the machine from public scanners.

**Comment by djmb** [comment]
@eric1234 - supporting IPv6 would be great, thank's for the great write-up of the issues! We'll definitely accept fixes for this.

If we don't include the square brackets in servers, but instead made sure to add them before passing them to SSHKit, would that solve the host comparison issue?

Regarding the setup, I think an `--ipv6` switch sounds like the way to go.

It would also be nice to get an integration tests that does an IPv6 based deployment if possible.

**Comment by jeromedalbert** [comment]
Thanks @eric1234 for this great deep dive. I stumbled upon the same issues a year ago but only got halfway through your solution before deciding to shell out an extra $0.5 to Hetzner and just get an IPv4 address. But I would love for Kamal to support IPv6 for sure!

> On one hand this seems "good enough" as it works. OTOH what downstream effects will be cause by overwriting to_s in a 3rd party package?

Maybe this concern could be addressed by using a ruby refinement. Still kind of a hack but at least it wouldn't leak elsewhere.

> It worked beautifully. I did a full setup, I was able to app exec, etc without error. It does mean that we are deciding to supporting ipv6 addresses without brackets while with brackets in probably more common (although the ssh program shows there are already exceptions). This seems fine to me, just want to make sure it's an explicit decision not one that we just made due to the impl convenience.

If implementation weren't an issue I think it would be nice to allow both syntaxes. Maybe a way to do it would be to strip any brackets when loading the Kamal yaml configuration, then put them back in like you're doing when calling SSHKit, although that would be quite hacky.

---

I think the best fix, if possible, is for SSHKit itself to parse any IPv6 form correctly in the first place, instead of making Kamal work around SSHKit's current limitations.

I submitted [an issue](https://github.com/capistrano/sshkit/issues/537) and a [PR](https://github.com/capistrano/sshkit/pull/538) for a potential fix. When I try that PR with Kamal, unbracketed and bracketed IPv6 addresses seem to work great. I didn't even need to do the Docker configuration part, it just worked for some reason, but I won't complain about it!
